docs/rmr.7.rst
.. This work is licensed under a Creative Commons Attribution 4.0 International License.
.. SPDX-License-Identifier: CC-BY-4.0
.. CAUTION: this document is generated from source in doc/src/rtd.
.. To make changes edit the source and recompile the document.
.. Do NOT make changes directly to .rst or .md files.

============================================================================================
Man Page: rmr
============================================================================================




RMR LIBRARY
===========



NAME
----

RMR -- Ric Message Router Library


DESCRIPTION
-----------

RMR is a library which provides a user application with the
ability to send and receive messages to/from other RMR based
applications without having to understand the underlying
messaging transport environment (e.g., SI95) and without
needing to know which other endpoint applications are
currently available and accepting messages. To do this, RMR
depends on a routing table generated by an external source.
This table is used to determine the destination endpoint of
each message sent by mapping the message type T (supplied by
the user application) to an endpoint entry. Once determined,
the message is sent directly to the endpoint. The user
application is unaware of which endpoint actually receives
the message, and in some cases whether that message was sent
to multiple applications.

RMR functions do provide for the ability to respond to the
specific source instance of a message allowing for either a
request response, or call response relationship when needed.


The Route Table
---------------

The library must be given a route table which maps message
types (integers) to endpoint groups such that each time a
message of type T is sent, the message is delivered to one
member of each group associated with T. For example, message
type 2 might route to two different groups where group A has
two members, worker1 and worker2, while group B has only one
member, logger1.

The route table consists of a start record, one or more table
entry records, and an end record. All table records contain
fields separated with vertical bars (|), and allow for
trailing comments with the standard shell comment symbol
(hash, #) provided that the start of the comment is separated
from the last token on the record by one or more spaces.
Leading and trailing white space in each field is ignored.
The route table supports two entry types: *rte* and *mse*.

A *rte* entry defines a message type, an optional sender
application, and the endpoint(s) which accept the indicated
message type. However, this format is deprecated and may be
removed in a future version. An example record appears next.

::

      rte | 1 | app10:4560


The second type of entry is *mse*. This entry defines a
message type, an optional sender application, a subscription
ID, and a collection of endpoints. An example record appears
next.

::

      mse | 1000,forwarder:43086 | 10 | app2:43086


It is the responsibility of the route table generator to know
which endpoints belong to which groups, and which groups
accept which message types. Once understood, the route table
generator publishes a table that is ingested by RMR and used
for mapping messages to end points.

The following is a simple route table which causes message
types 0 through 9 to be routed to specific applications:

::

  newrt|start
     mse|0|-1| %meid
     mse|1|-1|app10:4560,app11:4560
     mse|2|-1|app12:4560
     mse|3|-1|app14:4560
     mse|4|-1|app18:4560
     mse|5|-1|app01:4560
     mse|6|-1|app02:4560
     mse|7|-1|app03:4560
     mse|8|-1|app04:4560
     mse|9|-1|app05:4560
  newrt|end


The special endpoint "%meid" indicates that the message type
(0 in this case) is to be routed to the endpoint which has
been listed as the "owner" for the meid appearing in the
message. MEID ownership is communicated to RMR using the same
Route Table Manager interface and by supplying a "table" such
as the one below:

::

  meid_map | start
     mme_ar | control1 | meid000 meid001 meid002 meid003 meid004 meid005
     mme_ar | control2 | meid100 meid101 meid102 meid103
  meid_map | end | 2


This table indicates that the application (endpoint)
*control1* "owns" 6 MEIDs and *control2* owns 4. When message
type 0 is sent, the MEID in the message will be used to
select the endpoint via this table.

The MEID table will update the existing owner relationships,
and add new ones; it is necessary to send only the changes
with the add/replace (mme_ar) entries in the table. When
necessary, MEIDs can be deleted by adding an ``mme_del``
record to the table. The following example illustrates how
this might look:

::

  meid_map | start
     mme_ar | control1 | meid000 meid001 meid002 meid003 meid004 meid005
     mme_ar | control2 | meid100 meid101 meid102 meid103
     mme_del| meid200 meid401
  meid_map | end | 3



Route Table Syntax
------------------

The following illustrates the syntax for both types of route
table entries.


::

  newrt | start
  rte | <message-type>[,<sender-endpoint>] | <round-robin-grp>[;<round-robin-grp>]...
  mse | <message-type>[,<sender-endpoint>] | <sub-id> | <round-robin-grp>[;<round-robin-grp>]...
  newrt | end


A round robin group is one or more endpoints from which one
will be selected to receive the message. When multiple
endpoints are given in a group, they must be separated with a
comma. An endpoint is an IP address and port (e.g.
192.158.4.30:8219), or DNS name and port, of the application
that should receive the message type. If multiple round-robin
groups are given, they must be separated by a semicolon.


MEID Map Syntax
---------------

The MEID map is similar to the route table. Entries are used
to add or replace the ownership of one or more MEIDs (mme_ar)
or to delete one or more MEIDs (mme_del). The following is
the syntax for the MEID map.


::

  meid_map | start
  mme_ar | <owner-endpoint> | <meid> [<meid>...]
  mme_del | <meid> [<meid>...]
  meid_map | end | <count> | <md5sum>


The <count> on the end record indicates the number of mme_ar
and mme_del records which were sent; if the count does not
match the whole map is refused and dropped. The
<owner-endpoint> is the endpoint which should receive the
message when a message is routed based on the MEID it
contains. A MEID may be "owned" by only one endpoint, and if
supplied multiple times, the last observed relationship is
used. Each of the lists of MEIDs are blank separated.

The optional <md5sum> on the *end* record should be the
computed MD5 hash for all records which appear between the
start and and records. This allows for a tighter verification
that all data was received exactly as the route manager
transmitted them.


Environment
-----------

To enable configuration of the library behaviour outside of
direct user application control, RMR supports a number of
environment variables which provide information to the
library. The following is a list of the various environment
variables, what they control and the defaults which RMR uses
if undefined.

    .. list-table::
      :widths: auto
      :header-rows: 0
      :class: borderless

      * - **RMR_ASYNC_CONN**
        -
          Allows the async connection mode to be turned off (by setting
          the value to 0). When set to 1, or missing from the
          environment, RMR will invoke the connection interface in the
          transport mechanism using the non-blocking (async) mode. This
          will likely result in many "soft failures" (retry) until the
          connection is established, but allows the application to
          continue unimpeded should the connection be slow to set up.

      * - **RMR_BIND_IF**
        -
          This provides the interface that RMR will bind listen ports
          to, allowing for a single interface to be used rather than
          listening across all interfaces. This should be the IP
          address assigned to the interface that RMR should listen on,
          and if not defined RMR will listen on all interfaces.

      * - **RMR_CTL_PORT**
        -
          This variable defines the port that RMR should open for
          communications with Route Manager, and other RMR control
          applications. If not defined, the port 4561 is assumed.

          Previously, the ``RMR_RTG_SVC`` (route table generator
          service port) was used to define this port. However, a future
          version of Route Manager will require RMR to connect and
          request tables, thus that variable is now used to supply the
          Route Manager's well-known address and port.

          To maintain backwards compatibility with the older Route
          Manager versions, the presence of this variable in the
          environment will shift RMR's behaviour with respect to the
          default value used when ``RMR_RTG_SVC`` is **not** defined.

          When ``RMR_CTL_PORT`` is **defined:** RMR assumes that Route
          Manager requires RMR to connect and request table updates is
          made, and the default well-known address for Route manager is
          used (routemgr:4561).

          When ``RMR_CTL_PORT`` is **undefined:** RMR assumes that
          Route Manager will connect and push table updates, thus the
          default listen port (4561) is used.

          To avoid any possible misinterpretation and/or incorrect
          assumptions on the part of RMR, it is recommended that both
          the ``RMR_CTL_PORT`` and ``RMR_RTG_SVC`` be defined. In the
          case where both variables are defined, RMR will behave
          exactly as is communicated with the variable's values.

      * - **RMR_RTREQ_FREQ**
        -
          When RMR needs a new route table it will send a request once
          every ``n`` seconds. The default value for ``n`` is 5, but
          can be changed if this variable is set prior to invoking the
          process. Accepted values are between 1 and 300 inclusive.

      * - **RMR_RTG_SVC**
        -
          The value of this variable depends on the Route Manager in
          use.

          When the Route Manager is expecting to connect to an xAPP and
          push route tables, this variable must indicate the
          ``port`` which RMR should use to listen for these
          connections.

          When the Route Manager is expecting RMR to connect and
          request a table update during initialisation, the variable
          should be the ``host`` of the Route Manager process.

          The ``RMR_CTL_PORT`` variable (added with the support of
          sending table update requests to Route manager), controls the
          behaviour if this variable is not set. See the description of
          that variable for details.

      * - **RMR_HR_LOG**
        -
          By default RMR writes messages to standard error (incorrectly
          referred to as log messages) in human readable format. If
          this environment variable is set to 0, the format of standard
          error messages might be written in some format not easily
          read by humans. If missing, a value of 1 is assumed.

      * - **RMR_LOG_VLEVEL**
        -
          This is a numeric value which corresponds to the verbosity
          level used to limit messages written to standard error. The
          lower the number the less chatty RMR functions are during
          execution. The following is the current relationship between
          the value set on this variable and the messages written:


              .. list-table::
                :widths: auto
                :header-rows: 0
                :class: borderless

                * - **0**
                  -
                    Off; no messages of any sort are written.

                * - **1**
                  -

CHANGES

API and build change  and fix summaries. Doc correctsions
and/or changes are not mentioned here; see the commit messages.

2020 March 10; version 3.5.1
	Add missing health check message types.
	
2020 March 9; version 3.5.0
	Added new wormhole send function: rmr_wh_call().

2020 March 6; version 3.4.0
	Add new wormhole state function: rmr_wh_state().

2020 March 5; Version 3.3.1
	Correct several "bugs" identified by automatic code analysis.

2020 March 4; Version 3.3.0
	Add SI95 based unit testing
	Health check support binary added (reason for minor bump)

2020 February 29; Version 2.4.0
	Add consolidated testing under CMake
	Add support binary for health check (SI95 only)

2020 February 28; Version 2.3.6
	Fix bug in Rt. Mgr comm which prevented table ID from being
	sent on ack message (RIC-232).

2020 February 26; version 3.2.5
	Fix source address bug in SI95 receive/send funcitons.
	Fix threading issues involving session disconnection in SI95
	Remove unused SI95 status variable.

2020 February 24; version 3.2.4
	Fix meid bug (RIC-220) causing core dump.

2020 February 21; version 3.2.3
	Add meid routing support to the SI95 interface.

2020 February 20; version 3.2.2
	Fix receive thread related core dump (ring early unlock).

2020 February 19; version 3.2.1
	Added missing message types (E2-Setup)

2020 February 18; version 3.2.0
	Added support for new Route Manager and it's ability to accept
	a request for table update.

2020 February 14; version 3.1.3
	Fix bug in SIsend which was causing a core dump in some cases
	where the application attempted to send on a connection that
	had disconnected. (RIC-207).

2020 February 6; version 3.1.2
	Fix disconnection detection bug in interface to SI95.

2020 January 31; verison 3.1.1
	Allow route table thread logging to be completely disabled
	when logging is turned off.	

2020 January 26; verison 3.1.0
	First step to allowing the user programme to control messages
	written to standard error. Introduces the rmr_set_vlevel()
	function, and related environment variable.

2020 January 24; verison 3.0.5
	Fix bug in SI95 with receive buffer allocation.

2020 January 23; verison 3.0.4
	Fix bug in SI95 causing excessive CPU usage on poll.

2020 January 22; verison 3.0.3
	Enable thread support for multiple receive threads.

2020 January 21; verison 3.0.2
	Fix bug in SI95 (missing reallocate payload function).

2020 January 20; verison 3.0.1
	Enable support for dynamic route table updates via RMR session.

2020 January 16; version 3.0.0
	Introduce support for SI95 transport library to replace NNG.
	(RMR library versions will use leading odd numbers to avoid tag collisions
	 with the wrapper tags which will use even numbers.)

2019 December 9; version 1.13.1
	Correct documentation and missing rel-notes update for RTD.

2019 December 6; version 1.13.0
	Add ability to route messages based on the MEID in a message combined
	with the message type/subscription-ID.


2019 November 14; version 1.11.1 (Amber)
	Fix bug in payload reallocation function; correct length of payload
	was not always copied.

2019 November 13; version 1.12.1
	New message type constants added to support A1.

2019 November 4; version 1.11.0 (Amber)
	Version bump to move away from the 1.10.* to distinguish between
	release A and the trial.

2019 November 7; version 1.12.0
	Version cut to support continued development for next release
	preserving the 1.11.* versions for release 1 (Amber) and 
	related fixes.

2019 October 31; version 1.10.2
	Provide the means to increase the payload size of a received message
	without losing the data needed to use the rmr_rts_msg() funciton.

2019 October 21; version 1.10.1
	Fix to prevent null message buffer from being returned by the timeout
	receive function if the function is passed one to reuse.

2019 October 21; version 1.10.1
	Add periodic dump of send count info to stderr.

2019 September 27; version 1.9.0
	Python bindings added receive all queued function and corrected a unit test

2019 September 25; version 1.8.3
	Correct application level test issue causing timing problems during
	jenkins verification testing at command and merge

	Handle the NNG connection shutdown status which may now be
	generated when a connection throug a proxy is reset.

2019 September 25; version 1.8.2
	Correct bug in rmr_torcv_msg() when timeout set to zero (0).

2019 September 19; version 1.8.1
	Correct missing constant for wrappers.

2019 September 19; version 1.8.0
	New message types added:
		RAN_CONNECTED, RAN_RESTARTED, RAN_RECONFIGURED

2019 September 17; version 1.7.0
	Initial connection mode now defaults to asynchronous. Set RMR_ASYNC_CONN=0
	in the environment before rmr_init() is invoked to revert to
	synchronous first TCP connections. (Recovery connection attempts
	have always been asynchronous).

2019 September 3; version 1.6.0
	Fix bug in the rmr_rts_msg() function. If a return to sender message
	failed, the source IP address was not correctly adjusted and could
	cause the message to be "reflected" back to the sender on a retry.

	Added the ability to set the source "ID" via an environment var
	(RMR_SRC_ID).  When present in the environment, the string will
	be placed in to the message header as the source and thus be used
	by an application calling rmr_rts_smg() to return a response to
	the sender.  If this environment variable is not present, the host
	name (original behaviour) is used.

2019 August 26; version 1.4.0
	New message types were added.

2019 August 16; version 1.3.0
	New mesage types added.

2019 August 13; version 1.2.0 (API change, non-breaking)
	The function rmr_get_xact() was added to proide a convenient
	way to extract the transaction field from a message.

2019 August 8; version 1.1.0 (API change)
	This change should be backward compatable/non-breaking
	A new field has been added to the message buffer (rmr_mbuf_t).

src/bindings/rmr-python/rmr/rmr.py
# vim: expandtab ts=4 sw=4:
# ==================================================================================
#       Copyright (c) 2019-2020 Nokia
#       Copyright (c) 2018-2020 AT&T Intellectual Property.
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#          http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
# ==================================================================================
import uuid
import json
from ctypes import RTLD_GLOBAL, Structure, c_int, POINTER, c_char, c_char_p, c_void_p, memmove, cast
from ctypes import CDLL
from ctypes import create_string_buffer
from rmr.exceptions import BadBufferAllocation, MeidSizeOutOfRange, InitFailed

# https://docs.python.org/3.7/library/ctypes.html
# https://stackoverflow.com/questions/2327344/ctypes-loading-a-c-shared-library-that-has-dependencies/30845750#30845750
# make sure you do a set -x LD_LIBRARY_PATH /usr/local/lib/;

# even though we don't use these directly, they contain symbols we need
rmr_c_lib = CDLL("librmr_si.so", mode=RTLD_GLOBAL)


# Internal Helpers (not a part of public api)


_rmr_const = rmr_c_lib.rmr_get_consts
_rmr_const.argtypes = []
_rmr_const.restype = c_char_p


def _get_constants(cache={}):
    """
    Get or build needed constants from rmr
    TODO: are there constants that end user applications need?
    """
    if cache:
        return cache

    js = _rmr_const()  # read json string
    cache = json.loads(str(js.decode()))  # create constants value object as a hash
    return cache


def _get_mapping_dict(cache={}):
    """
    Get or build the state mapping dict

    RMR_OK              0   state is good
    RMR_ERR_BADARG      1   argument passd to function was unusable
    RMR_ERR_NOENDPT     2   send/call could not find an endpoint based on msg type
    RMR_ERR_EMPTY       3   msg received had no payload; attempt to send an empty message
    RMR_ERR_NOHDR       4   message didn't contain a valid header
    RMR_ERR_SENDFAILED  5   send failed; errno has nano reason
    RMR_ERR_CALLFAILED  6   unable to send call() message
    RMR_ERR_NOWHOPEN    7   no wormholes are open
    RMR_ERR_WHID        8   wormhole id was invalid
    RMR_ERR_OVERFLOW    9   operation would have busted through a buffer/field size
    RMR_ERR_RETRY       10  request (send/call/rts) failed, but caller should retry (EAGAIN for wrappers)
    RMR_ERR_RCVFAILED   11  receive failed (hard error)
    RMR_ERR_TIMEOUT     12  message processing call timed out
    RMR_ERR_UNSET       13  the message hasn't been populated with a transport buffer
    RMR_ERR_TRUNC       14  received message likely truncated
    RMR_ERR_INITFAILED  15  initialization of something (probably message) failed

    """
    if cache:
        return cache

    rmr_consts = _get_constants()
    for key in rmr_consts:  # build the state mapping dict
        if key[:7] in ["RMR_ERR", "RMR_OK"]:
            en = int(rmr_consts[key])
            cache[en] = key

    return cache


def _state_to_status(stateno):
    """
    Convert a msg state to status

    """
    sdict = _get_mapping_dict()
    return sdict.get(stateno, "UNKNOWN STATE")


_RCONST = _get_constants()


##############
# PUBLIC API
##############


# These constants are directly usable by importers of this library
# TODO: Are there others that will be useful?

RMR_MAX_RCV_BYTES = _RCONST["RMR_MAX_RCV_BYTES"]
RMRFL_MTCALL = _RCONST.get("RMRFL_MTCALL", 0x02)  # initialization flags
RMRFL_NONE = _RCONST.get("RMRFL_NONE", 0x0)
RMR_OK = _RCONST["RMR_OK"]  # useful state constants
RMR_ERR_TIMEOUT = _RCONST["RMR_ERR_TIMEOUT"]
RMR_ERR_RETRY = _RCONST["RMR_ERR_RETRY"]


class rmr_mbuf_t(Structure):
    """
    Reimplementation of rmr_mbuf_t which is in an unaccessible header file (src/common/include/rmr.h)

    | typedef struct {
    |    int     state;          // state of processing
    |    int     mtype;          // message type
    |    int     len;            // length of data in the payload (send or received)
    |    unsigned char* payload; // transported data
    |    unsigned char* xaction; // pointer to fixed length transaction id bytes
    |    int sub_id;             // subscription id
    |    int      tp_state;      // transport state (a.k.a errno)
    |
    | these things are off limits to the user application
    |
    |    void*   tp_buf;         // underlying transport allocated pointer (e.g. nng message)
    |    void*   header;         // internal message header (whole buffer: header+payload)
    |    unsigned char* id;      // if we need an ID in the message separate from the xaction id
    |    int flags;              // various MFL (private) flags as needed
    |    int alloc_len;          // the length of the allocated space (hdr+payload)
    | } rmr_mbuf_t;

    We do not include the fields we are not supposed to mess with

    RE PAYLOADs type below, see the documentation for c_char_p:
       class ctypes.c_char_p
           Represents the C char * datatype when it points to a zero-terminated string. For a general character pointer that may also point to binary data, POINTER(c_char) must be used. The constructor accepts an integer address, or a bytes object.
    """

    _fields_ = [
        ("state", c_int),
        ("mtype", c_int),
        ("len", c_int),
        (
            "payload",
            POINTER(c_char),
        ),  # according to th following the python bytes are already unsinged https://bytes.com/topic/python/answers/695078-ctypes-unsigned-char
        ("xaction", POINTER(c_char)),
        ("sub_id", c_int),
        ("tp_state", c_int),
    ]


# argtypes and restype are important: https://stackoverflow.com/questions/24377845/ctype-why-specify-argtypes


_rmr_init = rmr_c_lib.rmr_init
_rmr_init.argtypes = [c_char_p, c_int, c_int]
_rmr_init.restype = c_void_p


def rmr_init(uproto_port, max_msg_size, flags):
    """
    Refer to rmr C documentation for rmr_init
    extern void* rmr_init(char* uproto_port, int max_msg_size, int flags)

    This python function checks that the context is not None and raises
    an excption if it is.
    """
    mrc = _rmr_init(uproto_port, max_msg_size, flags)
    if mrc is None:
        raise InitFailed()
    return mrc


_rmr_ready = rmr_c_lib.rmr_ready
_rmr_ready.argtypes = [c_void_p]
_rmr_ready.restype = c_int


def rmr_ready(vctx):
    """
    Refer to rmr C documentation for rmr_ready
    extern int rmr_ready(void* vctx)
    """
    return _rmr_ready(vctx)


_rmr_close = rmr_c_lib.rmr_close
_rmr_close.argtypes = [c_void_p]


def rmr_close(vctx):
    """
    Refer to rmr C documentation for rmr_close
    extern void rmr_close(void* vctx)
    """
    return _rmr_close(vctx)


_rmr_set_stimeout = rmr_c_lib.rmr_set_stimeout
_rmr_set_stimeout.argtypes = [c_void_p, c_int]
_rmr_set_stimeout.restype = c_int


def rmr_set_stimeout(vctx, time):
    """
    Refer to the rmr C documentation for rmr_set_stimeout
    extern int rmr_set_stimeout(void* vctx, int time)
    """
    return _rmr_set_stimeout(vctx, time)


_rmr_alloc_msg = rmr_c_lib.rmr_alloc_msg
_rmr_alloc_msg.argtypes = [c_void_p, c_int]
_rmr_alloc_msg.restype = POINTER(rmr_mbuf_t)


def rmr_alloc_msg(
    vctx, size, payload=None, gen_transaction_id=False, mtype=None, meid=None, sub_id=None, fixed_transaction_id=None
):
    """
    Refer to the rmr C documentation for rmr_alloc_msg
    extern rmr_mbuf_t* rmr_alloc_msg(void* vctx, int size)
    TODO: on next API break, clean up transaction_id ugliness. Kept for now to preserve API.

    if payload is not None, attempts to set the payload
    if gen_transaction_id is True, it generates and sets a transaction id. Note, fixed_transaction_id supersedes this option
    if mtype is not None, sets the sbuf's message type
    if meid is not None, sets the sbuf's meid
    if sub_id is not None, sets the sbud's subscription id
    if fixed_transaction_id is set, it deterministically sets the transaction_id. This overrides the option gen_transation_id

    """
    sbuf = _rmr_alloc_msg(vctx, size)
    try:
        # make sure the alloc worked
        sbuf.contents

        # set specified fields
        if payload:
            set_payload_and_length(payload, sbuf)

        if fixed_transaction_id:
            set_transaction_id(sbuf, fixed_transaction_id)
        elif gen_transaction_id:
            generate_and_set_transaction_id(sbuf)

        if mtype:
            sbuf.contents.mtype = mtype

        if meid:
            rmr_set_meid(sbuf, meid)

        if sub_id:
            sbuf.contents.sub_id = sub_id

        return sbuf

    except ValueError:
        raise BadBufferAllocation


_rmr_realloc_payload = rmr_c_lib.rmr_realloc_payload
_rmr_realloc_payload.argtypes = [POINTER(rmr_mbuf_t), c_int, c_int, c_int]  # new_len, copy, clone
_rmr_realloc_payload.restype = POINTER(rmr_mbuf_t)


def rmr_realloc_payload(ptr_mbuf, new_len, copy=False, clone=False):
    """
    Refer to the rmr C documentation for rmr_realloc_payload().
    extern rmr_mbuf_t* rmr_realloc_payload(rmr_mbuf_t*, int, int, int)
    """
    return _rmr_realloc_payload(ptr_mbuf, new_len, copy, clone)


_rmr_free_msg = rmr_c_lib.rmr_free_msg
_rmr_free_msg.argtypes = [c_void_p]
_rmr_free_msg.restype = None


def rmr_free_msg(mbuf):
    """
    Refer to the rmr C documentation for rmr_free_msg
    extern void rmr_free_msg(rmr_mbuf_t* mbuf )
    """
    if mbuf is not None:
        _rmr_free_msg(mbuf)


_rmr_payload_size = rmr_c_lib.rmr_payload_size
_rmr_payload_size.argtypes = [POINTER(rmr_mbuf_t)]
_rmr_payload_size.restype = c_int


def rmr_payload_size(ptr_mbuf):
    """
    Refer to the rmr C documentation for rmr_payload_size
    extern int rmr_payload_size(rmr_mbuf_t* msg)
    """
    return _rmr_payload_size(ptr_mbuf)


"""
The following functions all seem to have the same interface
"""

_rmr_send_msg = rmr_c_lib.rmr_send_msg
_rmr_send_msg.argtypes = [c_void_p, POINTER(rmr_mbuf_t)]
_rmr_send_msg.restype = POINTER(rmr_mbuf_t)


def rmr_send_msg(vctx, ptr_mbuf):
    """
    Refer to the rmr C documentation for rmr_send_msg
    extern rmr_mbuf_t* rmr_send_msg(void* vctx, rmr_mbuf_t* msg)
    """
    return _rmr_send_msg(vctx, ptr_mbuf)


# TODO: the old message (Send param) is actually optional, but I don't know how to specify that in Ctypes.
_rmr_rcv_msg = rmr_c_lib.rmr_rcv_msg
_rmr_rcv_msg.argtypes = [c_void_p, POINTER(rmr_mbuf_t)]
_rmr_rcv_msg.restype = POINTER(rmr_mbuf_t)


def rmr_rcv_msg(vctx, ptr_mbuf):
    """
    Refer to the rmr C documentation for rmr_rcv_msg
    extern rmr_mbuf_t* rmr_rcv_msg(void* vctx, rmr_mbuf_t* old_msg)
    """
    return _rmr_rcv_msg(vctx, ptr_mbuf)


_rmr_torcv_msg = rmr_c_lib.rmr_torcv_msg
_rmr_torcv_msg.argtypes = [c_void_p, POINTER(rmr_mbuf_t), c_int]
_rmr_torcv_msg.restype = POINTER(rmr_mbuf_t)


def rmr_torcv_msg(vctx, ptr_mbuf, ms_to):
    """
    Refer to the rmr C documentation for rmr_torcv_msg
    extern rmr_mbuf_t* rmr_torcv_msg(void* vctx, rmr_mbuf_t* old_msg, int ms_to)
    """
    return _rmr_torcv_msg(vctx, ptr_mbuf, ms_to)


_rmr_rts_msg = rmr_c_lib.rmr_rts_msg
_rmr_rts_msg.argtypes = [c_void_p, POINTER(rmr_mbuf_t)]
_rmr_rts_msg.restype = POINTER(rmr_mbuf_t)


def rmr_rts_msg(vctx, ptr_mbuf, payload=None, mtype=None):
    """
    Refer to the rmr C documentation for rmr_rts_msg
    extern rmr_mbuf_t*  rmr_rts_msg(void* vctx, rmr_mbuf_t* msg)

    additional features beyond c-rmr:
        if payload is not None, attempts to set the payload
        if mtype is not None, sets the sbuf's message type
    """

    if payload:
        set_payload_and_length(payload, ptr_mbuf)

    if mtype:
        ptr_mbuf.contents.mtype = mtype

    return _rmr_rts_msg(vctx, ptr_mbuf)


_rmr_call = rmr_c_lib.rmr_call
_rmr_call.argtypes = [c_void_p, POINTER(rmr_mbuf_t)]
_rmr_call.restype = POINTER(rmr_mbuf_t)


def rmr_call(vctx, ptr_mbuf):
    """
    Refer to the rmr C documentation for rmr_call
    extern rmr_mbuf_t* rmr_call(void* vctx, rmr_mbuf_t* msg)
    """
    return _rmr_call(vctx, ptr_mbuf)


_rmr_bytes2meid = rmr_c_lib.rmr_bytes2meid
_rmr_bytes2meid.argtypes = [POINTER(rmr_mbuf_t), c_char_p, c_int]
_rmr_bytes2meid.restype = c_int


def rmr_set_meid(ptr_mbuf, byte_str):
    """
    Refer to the rmr C documentation for rmr_bytes2meid
    extern int rmr_bytes2meid(rmr_mbuf_t* mbuf, unsigned char const* src, int len);

    Caution:  the meid length supported in an RMR message is 32 bytes, but C applications
    expect this to be a nil terminated string and thus only 31 bytes are actually available.

    Raises: exceptions.MeidSizeOutOfRang
    """
    max = _get_constants().get("RMR_MAX_MEID", 32)
    if len(byte_str) >= max:
        raise MeidSizeOutOfRange

    return _rmr_bytes2meid(ptr_mbuf, byte_str, len(byte_str))


# CAUTION:  Some of the C functions expect a mutable buffer to copy the bytes into;
#           if there is a get_* function below, use it to set up and return the
#           buffer properly.

# extern unsigned char*  rmr_get_meid(rmr_mbuf_t* mbuf, unsigned char* dest);
# we don't provide direct access to this function (unless it is asked for) because it is not really useful to provide your own buffer.
# Rather, rmr_get_meid does this for you, and just returns the string.
_rmr_get_meid = rmr_c_lib.rmr_get_meid
_rmr_get_meid.argtypes = [POINTER(rmr_mbuf_t), c_char_p]
_rmr_get_meid.restype = c_char_p


def rmr_get_meid(ptr_mbuf):
    """
    Get the managed equipment ID (meid) from the message header.

    Parameters
    ----------
    ptr_mbuf: ctypes c_void_p
        Pointer to an rmr message buffer

    Returns
    -------
    string:
        meid
    """
    sz = _get_constants().get("RMR_MAX_MEID", 32)  # size for buffer to fill
    buf = create_string_buffer(sz)
    _rmr_get_meid(ptr_mbuf, buf)
    return buf.value


_rmr_get_src = rmr_c_lib.rmr_get_src
_rmr_get_src.argtypes = [POINTER(rmr_mbuf_t), c_char_p]
_rmr_get_src.restype = c_char_p


def rmr_get_src(ptr_mbuf, dest):
    """
    Refer to the rmr C documentation for rmr_get_src
    extern unsigned char*  rmr_get_src(rmr_mbuf_t* mbuf, unsigned char* dest);
    """
    return _rmr_get_src(ptr_mbuf, dest)


# Methods that exist ONLY in rmr-python, and are not wrapped methods
# In hindsight, I wish i put these in a seperate module, but leaving this here to prevent api breakage.


def get_payload(ptr_mbuf):
    """
    Given a rmr_buf_t*, get it's binary payload as a bytes object

    Parameters
    ----------
    ptr_mbuf: ctypes c_void_p
        Pointer to an rmr message buffer

    Returns
    -------
    bytes:
        the message payload
    """
    # Logic came from the answer here: https://stackoverflow.com/questions/55103298/python-ctypes-read-pointerc-char-in-python
    sz = ptr_mbuf.contents.len
    CharArr = c_char * sz
    return CharArr(*ptr_mbuf.contents.payload[:sz]).raw


def get_xaction(ptr_mbuf):
    """
    given a rmr_buf_t*, get it's transaction id

    Parameters
    ----------
    ptr_mbuf: ctypes c_void_p
        Pointer to an rmr message buffer

    Returns
    -------
    bytes:
        the transaction id
    """
    val = cast(ptr_mbuf.contents.xaction, c_char_p).value
    sz = _get_constants().get("RMR_MAX_XID", 0)
    return val[:sz]


def message_summary(ptr_mbuf):
    """
    Returns a dict that contains the fields of a message

    Parameters
    ----------
    ptr_mbuf: ctypes c_void_p
        Pointer to an rmr message buffer

    Returns
    -------
    dict:
        dict message summary
    """
    return {
        "payload": get_payload(ptr_mbuf) if ptr_mbuf.contents.state == RMR_OK else None,
        "payload length": ptr_mbuf.contents.len,
        "message type": ptr_mbuf.contents.mtype,
        "subscription id": ptr_mbuf.contents.sub_id,
        "transaction id": get_xaction(ptr_mbuf),
        "message state": ptr_mbuf.contents.state,
        "message status": _state_to_status(ptr_mbuf.contents.state),
        "payload max size": rmr_payload_size(ptr_mbuf),
        "meid": rmr_get_meid(ptr_mbuf),
        "message source": get_src(ptr_mbuf),
        "errno": ptr_mbuf.contents.tp_state,
    }


def set_payload_and_length(byte_str, ptr_mbuf):
    """
    | Set an rmr payload and content length
    | In place method, no return

    Parameters
    ----------
    byte_str: bytes
        the bytes to set the payload to
    ptr_mbuf: ctypes c_void_p
        Pointer to an rmr message buffer
    """
    if rmr_payload_size(ptr_mbuf) < len(byte_str):  # existing message payload too small
        ptr_mbuf = rmr_realloc_payload(ptr_mbuf, len(byte_str), True)

    memmove(ptr_mbuf.contents.payload, byte_str, len(byte_str))
    ptr_mbuf.contents.len = len(byte_str)


def generate_and_set_transaction_id(ptr_mbuf):
    """
    Generate a UUID and Set an rmr transaction id to it

    Parameters
    ----------
    ptr_mbuf: ctypes c_void_p
        Pointer to an rmr message buffer
    """
    set_transaction_id(ptr_mbuf, uuid.uuid1().hex.encode("utf-8"))


def set_transaction_id(ptr_mbuf, tid_bytes):
    """
    Set an rmr transaction id
    TODO: on next API break, merge these two functions. Not done now to preserve API.

    Parameters
    ----------
    ptr_mbuf: ctypes c_void_p
        Pointer to an rmr message buffer
    tid_bytes: bytes
        bytes of the desired transaction id
    """
    sz = _get_constants().get("RMR_MAX_XID", 0)
    memmove(ptr_mbuf.contents.xaction, tid_bytes, sz)


def get_src(ptr_mbuf):
    """
    Get the message source (likely host:port)

    Parameters
    ----------
    ptr_mbuf: ctypes c_void_p
        Pointer to an rmr message buffer

    Returns
    -------
    string:
        message source
    """
    sz = _get_constants().get("RMR_MAX_SRC", 64)  # size to fill
    buf = create_string_buffer(sz)
    rmr_get_src(ptr_mbuf, buf)
    return buf.value.decode()
